internal/worldstate/leveldb/snapshot.go

```diff
@@ -0,0 +1,113 @@
+package leveldb
+
+import (
+	"sync"
+
+	"github.com/IBM-Blockchain/bcdb-server/internal/worldstate"
+	"github.com/IBM-Blockchain/bcdb-server/pkg/types"
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+type Snapshots struct {
+	dbSnap map[string]*leveldb.Snapshot
+	sync.RWMutex
+}
+
+func (l *LevelDB) GetSnapshot(dbNames []string) (worldstate.DBSnapshot, error) {
+	l.dbsList.RLock()
+	defer l.dbsList.RUnlock()
+
+	snap := &Snapshots{
+		dbSnap: make(map[string]*leveldb.Snapshot),
+	}
+
+	for _, dbName := range dbNames {
+		db, ok := l.dbs[dbName]
+		if !ok {
+			return nil, &DBNotFoundErr{
+				dbName: dbName,
+			}
+		}
+
+		db.mu.RLock()
+		defer db.mu.RUnlock()
+
+		s, err := db.file.GetSnapshot()
+		if err != nil {
+			return nil, err
+		}
+
+		snap.dbSnap[dbName] = s
+	}
+
+	return snap, nil
+}
+
+func (s *Snapshots) Get(dbName, key string) ([]byte, *types.Metadata, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	lSnap, ok := s.dbSnap[dbName]
+	if !ok {
+		return nil, nil, errors.New(dbName + " is needed to fetch the index definiton and is not snapshotted")
+	}
+
+	dbval, err := lSnap.Get([]byte(key), &opt.ReadOptions{})
+	if err == leveldb.ErrNotFound {
+		return nil, nil, nil
+	}
+	if err != nil {
+		return nil, nil, errors.WithMessagef(err, "failed to retrieve leveldb key [%s] from database %s", key, dbName)
```

Comment:
"failed to retrieve leveldb key [%s] from database %s snapshot"

Reply:
Done